libirods_rule_engine_plugin-audit_amqp.cpp

```diff
@@ -1,413 +1,544 @@
+#include <proton/messaging_handler.hpp>
+#include <proton/tracker.hpp>
+#include <proton/sender.hpp>
+
+#if __cpp_lib_chrono >= 201907
+// use utc_clock if it's implemented (for leap seconds)
+using ts_clock = std::chrono::utc_clock;
+#else
+// fallback to system_clock
+using ts_clock = std::chrono::system_clock;
+#endif
+
+using log_re = irods::experimental::log::rule_engine;
+
+// NOLINTBEGIN(cert-err58-cpp, cppcoreguidelines-avoid-non-const-global-variables)
+static const std::string rule_engine_name("audit_amqp");
+static const std::string default_pep_regex_to_match("audit_.*");
+static const std::string default_amqp_topic("irods_audit_messages");
+static const std::string default_amqp_location("localhost:5672");
+static const std::string default_amqp_user("");
+static const std::string default_amqp_password("");
+static const std::string default_amqp_options("");
+
+static std::string audit_pep_regex_to_match(default_pep_regex_to_match);
+static std::string audit_amqp_topic(default_amqp_topic);
+static std::string audit_amqp_location(default_amqp_location);
+static std::string audit_amqp_user(default_amqp_user);
+static std::string audit_amqp_password(default_amqp_password);
+static std::string audit_amqp_options(default_amqp_options);
+
+static std::mutex audit_plugin_mutex;
+// NOLINTEND(cert-err58-cpp, cppcoreguidelines-avoid-non-const-global-variables)
+
+// See qpid-cpp docs (https://qpid.apache.org/releases/qpid-proton-0.37.0/proton/cpp/api/simple_send_8cpp-example.html)
+class send_handler : public proton::messaging_handler
+{
+  private:
+	const std::string& amqp_location;
+	const std::string& amqp_topic;
+	const std::string& user;
+	const std::string& password;
+	const std::string& message;
+	proton::sender sender;
+
+  public:
+	// NOLINTNEXTLINE(bugprone-easily-swappable-parameters)
+	send_handler(
+		const std::string& _message,
+		const std::string& _location,
+		const std::string& _topic,
+		const std::string& _user,
+		const std::string& _password)
+		: amqp_location(_location)
+		, amqp_topic(_topic)
+		, user(_user)
+		, password(_password)
+		, message(_message)
+	{
+	}
+
+	void on_container_start(proton::container& container) override
+	{
+		proton::connection_options conn_opts;
+		if (!user.empty()) {
+			conn_opts.user(user);
+		}
+		if (!password.empty()) {
+			conn_opts.password(password);
+		}
+		sender = container.open_sender(amqp_location + '/' + amqp_topic, conn_opts);
+	}
+
+	void on_tracker_accept(proton::tracker& tracker) override
+	{
+		tracker.connection().close(); // we're only sending one message
+		                              // so we don't care about the credit system
+		                              // or tracking confirmed messages
+	}
+
+	void on_sendable(proton::sender& _sender) override
+	{
+		proton::message m(message); // NOLINT(readability-identifier-length)
+		m.content_type("application/json");
+		_sender.send(m);
+	}
+}; // class send_handler
+
+static BOOST_FORCEINLINE void
```

Comment:
The way to do this might differ between compilers. Boost provides a macro, so I used it.

Reply:
Sure. My question is more about what sparked you to add it? Was it there before or did you notice something that made you investigate it?

Reply:
When I started working on this, these functions were neither static nor inline, even though they were not used outside of this translation unit.
Since the standard use case seems to be to have this plugin fire off for every rule that gets executed, I think inlining makes sense for the performance benefit.